juju/model.py

```diff
@@ -0,0 +1,210 @@
+import json
+from collections.abc import Mapping
+from functools import lru_cache
+from subprocess import run, PIPE
+
+
+class Model:
+    def __init__(self, app_name, unit_name, relation_names, backend=None):
+        self.app = Application(app_name, self._lazy_load_peer_units)
+        self.unit = Unit(unit_name)
+        self._backend = backend or ModelBackend()
+        self.relations = RelationMap(relation_names, self.app, self.unit, self._backend)
+
+    def relation(self, relation_name):
+        """Return the first Relation object for the named relation, or None.
+
+        This is a convenience method for the case where only a single related
+        app is supported. If there are more than one remote apps on the relation,
+        it will raise a TooManyRelatedApps model error. If no there are no remote
+        apps on the relation, it will return None. Otherwise, it is equivalent
+        to self.relations[relation_name][0].
+        """
+        num_related = len(self.relations[relation_name])
+        if num_related > 1:
+            raise TooManyRelatedApps(relation_name, num_related, 1)
+        elif num_related == 0:
+            return None
+        else:
+            return self.relations[relation_name][0]
+
+    def _lazy_load_peer_units(self):
+        # In lieu of an implicit peer relation, goal-state seems to be the only way to find out
+        # what our peer units are. The information it returns is rather limited, but is enough
+        # for our purposes here.
+        return [Unit(unit_name) for unit_name in self._backend.goal_state()['units']]
+
+
+class Application:
+    def __init__(self, name, units):
+        self.name = name
+        self._units = units
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+    @property
+    def units(self):
+        # This property allows for the list of units to be lazy-loaded by passing in a function instead of a list.
+        if callable(self._units):
+            self._units = self._units()
```

Comment:
The loading logic still feels somewhat "abstract" considering what we're going after. I'll make this comment more clear and tangible in the rest of the review.